File: lint.py
import os
import re
import json

def check_html():
    issues = []
    html_files = [f for f in os.listdir('.') if f.endswith('.html')]
    for file in html_files:
        with open(file, 'r', encoding='utf-8') as f:
            content = f.read()
            
            # Check for empty href
            empty_hrefs = re.findall(r'href=["\']\s*["\']', content)
            if empty_hrefs:
                issues.append(f"{file}: Found empty hrefs: {len(empty_hrefs)}")
                
            # Check for multiple main tags
            mains = re.findall(r'<main', content)
            if len(mains) > 1:
                issues.append(f"{file}: Multiple <main> tags found: {len(mains)}")
                
            # Check for duplicate IDs
            ids = re.findall(r'id=["\']([^"\']+)["\']', content)
            import collections
            duplicates = [item for item, count in collections.Counter(ids).items() if count > 1]
            if duplicates:
                issues.append(f"{file}: Duplicate IDs found: {', '.join(duplicates)}")
                
            # Basic tag matching
            tags_to_check = ['div', 'span', 'section', 'a', 'p', 'h1', 'h2', 'h3']
            for tag in tags_to_check:
                opens = len(re.findall(f'<{tag}[>\s]', content))
                closes = len(re.findall(f'</{tag}>', content))
                if opens != closes:
                    issues.append(f"{file}: Mismatched <{tag}> tags (open: {opens}, close: {closes})")
                    
    return issues

def check_json():
    issues = []
    json_files = [f for f in os.listdir('.') if f.endswith('.json')]
    for file in json_files:
        try:
            with open(file, 'r', encoding='utf-8') as f:
                json.load(f)
        except Exception as e:
            issues.append(f"{file}: Invalid JSON - {str(e)}")
    return issues

html_issues = check_html()
json_issues = check_json()

print("--- HTML ISSUES ---")
for issue in html_issues:
    print(issue)
print("\n--- JSON ISSUES ---")
for issue in json_issues:
    print(issue)

File: lint.js
const fs = require('fs');
const path = require('path');

function checkHtml() {
    const issues = [];
    const files = fs.readdirSync('.').filter(f => f.endsWith('.html'));
    
    files.forEach(file => {
        const content = fs.readFileSync(file, 'utf8');
        
        // Empty hrefs
        const emptyHrefs = content.match(/href=["']\s*["']/g);
        if (emptyHrefs) issues.push(`${file}: Found empty hrefs: ${emptyHrefs.length}`);
        
        // Multiple main tags
        const mains = content.match(/<main/g);
        if (mains && mains.length > 1) issues.push(`${file}: Multiple <main> tags found: ${mains.length}`);
        
        // Duplicate IDs
        const ids = [];
        let match;
        const idRegex = /id=["']([^"']+)["']/g;
        while ((match = idRegex.exec(content)) !== null) {
            ids.push(match[1]);
        }
        
        const counts = {};
        ids.forEach(id => counts[id] = (counts[id] || 0) + 1);
        const dups = Object.keys(counts).filter(id => counts[id] > 1);
        if (dups.length > 0) issues.push(`${file}: Duplicate IDs found: ${dups.join(', ')}`);
        
        // Tag matching
        ['div', 'span', 'section', 'a', 'p', 'h1', 'h2', 'h3'].forEach(tag => {
            const openMatch = content.match(new RegExp(`<${tag}[>\\s]`, 'g'));
            const closeMatch = content.match(new RegExp(`</${tag}>`, 'g'));
            const opens = openMatch ? openMatch.length : 0;
            const closes = closeMatch ? closeMatch.length : 0;
            if (opens !== closes && tag !== 'input' && tag !== 'img' && tag !== 'br' && tag !== 'hr') {
                issues.push(`${file}: Mismatched <${tag}> tags (open: ${opens}, close: ${closes})`);
            }
        });
    });
    return issues;
}

function checkJson() {
    const issues = [];
    const files = fs.readdirSync('.').filter(f => f.endsWith('.json'));
    
    files.forEach(file => {
        try {
            JSON.parse(fs.readFileSync(file, 'utf8'));
        } catch (e) {
            issues.push(`${file}: Invalid JSON - ${e.message}`);
        }
    });
    return issues;
}

const htmlIssues = checkHtml();
const jsonIssues = checkJson();

console.log("--- HTML ISSUES ---");
htmlIssues.forEach(i => console.log(i));
console.log("\n--- JSON ISSUES ---");
jsonIssues.forEach(i => console.log(i));
